### PR TITLE
public api should be protected by oauth token

### PR DESCRIPTION
**Pull request by qibobo**
[#150627701]

### CODE REVIEW

File: api/lib/oauth/oauth.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+          json: true,
+          timeout: 10000,
+          headers: {
+            "Authorization": userToken,
+            "Content-Type": "application/json"
+          }
+        };
+        request(options, function(error1, response, body) {
+          if (error1) {
+            logger.error("Error occurred during get space by userId and appId", { "appId": appId, "userId": userId, "error": error1 });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Failed to get space with userId and AppId during permission check

File: api/lib/oauth/oauth.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+              } else {
+                logger.info("Check user authorization failed", { "appId": appId, "userId": userId });
+                callback(null, false);
+              }
+
+            } else {
+              var errorObj = {
+                "statusCode": response.statusCode
+              };
+              logger.error("Error occurred during get space by userId and appId", { "appId": appId, "userId": userId, "error": errorObj });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Failed to get space with userId and AppId during permission check

File: api/lib/oauth/oauth.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        };
+        request(options, function(error1, response, body) {
+          if (error1) {
+            logger.error("Error occurred during get space by userId and appId", { "appId": appId, "userId": userId, "error": error1 });
+            callback(error1, null);
+          } else {
+            if (response.statusCode == HttpStatus.OK) {
+              var totalResults = body.total_results;
+              if (totalResults > 0) {
+                logger.info("Check user authorization successfully", { "appId": appId, "userId": userId })
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Permission check passed

File: api/lib/oauth/oauth.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                callback(null, true);
+              } else {
+                logger.info("Check user authorization failed", { "appId": appId, "userId": userId });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: permission check  not passed

File: api/lib/oauth/oauth.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  function getCloudControllerEndpoint(req) {
+    return settings.cfApi;;
+  };
+
+  function getCloudFoundryInfo(req, callback) {
+    var ccEndpoint = getCloudControllerEndpoint(req);
+    var options = {
+      url: ccEndpoint + "/v2/info",
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: do we need to normalize ccEndpoint?

File: api/lib/oauth/oauth.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      method: "GET",
+      json: true,
+      timeout: 10000,
+
+    };
+    request(options, function(error, response, body) {
+      if (error) {
+        logger.error("Error occurred during get cloud foundry information ", { "info url": options.url, "error": error });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: failed to get Cloud Foundry API information

File: api/lib/oauth/oauth.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        error.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+        callback(error, null);
+      } else if (response.statusCode === HttpStatus.OK) {
+        logger.info("Get cound foundry information successfully", { "info url": options.url });
+        callback(null, { "statusCode": HttpStatus.OK, "body": body });
+      } else {
+        var errorObj = {
+          "statusCode": response.statusCode
+        };
+        logger.error("Error occurred during get cloud foundry information ", { "info url": options.url, "error": errorObj });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Failed to get Cloud Foundry API information

File: api/lib/oauth/oauth.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        callback(errorObj, null);
+      }
+    });
+  }
+
+
+
+  function getUserInfo(req, callback) {
+    getCloudFoundryInfo(req, function(error, responseBody) {
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: why getting uaa endpoint every time?

File: api/lib/oauth/oauth.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+          json: true,
+          timeout: 10000,
+          headers: {
+            "Authorization": userToken,
+            "Content-Type": "application/json"
+          }
+        };
+        request(options, function(error1, response, body) {
+          if (error1) {
+            logger.error("Error occurred during get userinfo", { "userToken": userToken, "http-options": options, "error": error1 });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Failed to get user info from UAA

File: api/lib/oauth/oauth.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+            error1.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+            callback(error1, null);
+          } else {
+            if (response.statusCode == HttpStatus.OK) {
+              callback(null, body.user_id);
+            } else {
+              var errorObj = {
+                "statusCode": response.statusCode
+              };
+              logger.error("Error occurred during get userinfo", { "userToken": userToken, "http-options": options, "error": errorObj });
```
Comment (boyang9527) [inline api/lib/oauth/oauth.js]: Failed to get user info from UAA

File: api/test/unit/oauth/oauth.test.js [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+      });
+      it("should return false", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(false);
+          expect(error).to.equal(null);
+          done();
+        });
+      });
+    });
+    context("cloudController info endpoint is down", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Cloud Controller  API endpoint is not available

File: api/test/unit/oauth/oauth.test.js [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+      it("should return error", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error.statusCode).to.equal(HttpStatus.INTERNAL_SERVER_ERROR);
+          done();
+        });
+      });
+    });
+
+    context("cloudController info endpoint returns error", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Cloud Controller /v2/info returns 400

File: api/test/unit/oauth/oauth.test.js [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+          expect(result).to.equal(null);
+          expect(error).to.deep.equal({
+            "statusCode": HttpStatus.BAD_REQUEST
+          });
+          done();
+        });
+      });
+    });
+
+    context("authorizationEndpoint is down", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Authorization endpoint is not available

File: api/test/unit/oauth/oauth.test.js [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+      it("should return error", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error.statusCode).to.equal(HttpStatus.INTERNAL_SERVER_ERROR);
+          done();
+        });
+      });
+    });
+
+    context("authorizationEndpoint returns error", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Authorization endpoint returns 400

File: api/test/unit/oauth/oauth.test.js [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+          expect(result).to.equal(null);
+          expect(error).to.deep.equal({
+            "statusCode": HttpStatus.BAD_REQUEST
+          });
+          done();
+        });
+      });
+    });
+
+    context("check user space endpoint is down", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Cloud Controller user spaces API is not available

File: api/test/unit/oauth/oauth.test.js [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+      it("should return error", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error.statusCode).to.equal(HttpStatus.INTERNAL_SERVER_ERROR);
+          done();
+        });
+      });
+    });
+
+    context("check user space endpoint returns error", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: Cloud Controller user spaces api returns 400

File: api/test/unit/oauth/oauth.test.js [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+          expect(result).to.equal(null);
+          expect(error).to.deep.equal({
+            "statusCode": HttpStatus.BAD_REQUEST
+          });
+          done();
+        });
+      });
+    });
+
+    context("check user space endpoint returns error", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: duplicate ?

File: src/integration/integration_api_metricscollector_test.go [hunk 18]
```diff
@@ -53,6 +54,22 @@ var _ = Describe("Integration_Api_MetricsCollector", func() {
 		stopMetricsCollector()
 	})
 	Describe("Get metrics", func() {
+
+		Context("Check user authorization failed", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_metricscollector_test.go]: check permission not passed 

you may add another test case for  cloud controller/uaa  api not available

File: src/integration/integration_api_scalingengine_test.go [hunk 19]
```diff
@@ -47,6 +48,22 @@ var _ = Describe("Integration_Api_ScalingEngine", func() {
 		stopScalingEngine()
 	})
 	Describe("Get scaling histories", func() {
+		Context("Check user authorization failed", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_scalingengine_test.go]: same for metrics

File: src/integration/integration_api_scheduler_test.go [hunk 20]
```diff
@@ -38,6 +41,37 @@ var _ = Describe("Integration_Api_Scheduler", func() {
 		stopScheduler(schedulerProcess)
 	})
 
+	Context("Check user authorization failed", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_scheduler_test.go]: same for metrics

**Comment by qibobo** [comment]
@boyang9527 updated.

File: api/lib/oauth/oauth.js [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+    });
+
+  }
+
+  function getCloudControllerEndpoint(req) {
+    return settings.cfApi;;
+  };
+
+  function getCloudFoundryInfo(req, callback) {
```
Comment (cdlliuy) [inline api/lib/oauth/oauth.js]: Is that unnecessary to pass-through "req" here?  the cc endpoint is set through settings as defined in func getCloudControllerEndpoint, not generated from "req"

File: api/lib/oauth/oauth.js [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+        }
+      });
+    } else {
+      getCloudFoundryInfo(req, function(error, responseBody) {
+        if (error) {
+          callback(error, null);
+        } else {
+          var userToken = req.header("Authorization");
+          obj.authorizationEndpoint = responseBody.body.authorization_endpoint;
+          var options = {
```
Comment (cdlliuy) [inline api/lib/oauth/oauth.js]: These codes are duplicated to the implementations of L110-L133.  
How about do checking on obj.authorizationEndpoint first. If not exist, then set it with getCloudFoundryInfo().  Once obj.authorizationEndpoint is set correctly,  invoke /usrinfo directly

File: api/app.js [hunk 23]
```diff
@@ -90,6 +92,41 @@ module.exports = function(configFilePath) {
     res.status(HttpStatus.BAD_REQUEST).json(errorResponse);
   });
 
+  var publicApp = express();
+  publicApp.use(bodyParser.json());
+  publicApp.use(bodyParser.urlencoded({ extended: false }));
+  publicApp.use(function(req, res, next){
+    oauth.checkUserAuthorization(req,function(error,isSpaceDeveloper){
+      if(error){
+        res.status(HttpStatus.BAD_REQUEST).send({});
```
Comment (cdlliuy) [inline api/app.js]: According to "checkUserAuthorization",  the possible error could be 
1. error when connect to cc or uaa
2. invalid user token

It is incorrect to return 400 bad request in these condition

File: api/test/unit/config/settings_test.js [hunk 24]
```diff
@@ -12,6 +12,7 @@ describe('config setting Test Suite', function() {
     defaultConfig = {
       "port": 8080,
       "publicPort": 8081,
+      "cfApi": "http://api.bosh-lite.com",
```
Comment (cdlliuy) [inline api/test/unit/config/settings_test.js]: in setting.js. the default setting is https

File: api/test/unit/oauth/oauth.test.js [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+      it("should return error", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error.statusCode).to.equal(HttpStatus.INTERNAL_SERVER_ERROR);
+          done();
+        });
+      });
+    });
+
+    context("Cloud Controller /v2/info returns 400", function() {
```
Comment (cdlliuy) [inline api/test/unit/oauth/oauth.test.js]: how about return 503 with cc /v2/info?

File: src/integration/integration_api_scalingengine_test.go [hunk 26]
```diff
@@ -47,6 +48,52 @@ var _ = Describe("Integration_Api_ScalingEngine", func() {
...
+				fakeCCNOAAUAA.AllowUnhandledRequests = true
+				fakeCCNOAAUAA.RouteToHandler("GET", "/v2/info", ghttp.RespondWithJSONEncoded(http.StatusOK,
+					cf.Endpoints{
+						AuthEndpoint:    fakeCCNOAAUAA.URL(),
+						DopplerEndpoint: strings.Replace(fakeCCNOAAUAA.URL(), "http", "ws", 1),
+					}))
+				parameters = map[string]string{"start-time": "1111", "end-time": "9999", "order": "desc", "page": "1", "results-per-page": "5"}
+			})
+			It("should error when access public api", func() {
+				By("check public api")
```
Comment (cdlliuy) [inline src/integration/integration_api_scalingengine_test.go]: as explained in app.js,  it is incorrect to return 400 bad request when  "cc /uaa unavailable"

File: api/lib/oauth/oauth.js [hunk 27]
```diff
@@ -16,7 +16,7 @@ module.exports = function(configFilePath) {
     var appId = reqPath.substring(startIndex, endIndex);
     if (!appId) {
       logger.error("Failed to get appId");
-      callback({ "message": "Failed to get appId" }, null);
+      callback({ "statusCode": HttpStatus.UNAUTHORIZED, "message": "Failed to get appId" }, null);
```
Comment (cdlliuy) [inline api/lib/oauth/oauth.js]: If "appId" is not passed through in the request,  i.e. /v1/apps/policy/,  it should be 404 not found ..

File: api/lib/oauth/oauth.js [hunk 28]
```diff
@@ -103,71 +104,47 @@ module.exports = function(configFilePath) {
...
+          };
+          logger.error("Failed to get user info from UAA", { "userToken": userToken, "http-options": options, "error": errorObj });
+          callback(errorObj, null);
+        }
+      }
+    });
+  }
 
   function getUserInfo(req, callback) {
     if (obj.authorizationEndpoint) {
```
Comment (cdlliuy) [inline api/lib/oauth/oauth.js]: how about
 if (!obj.authorizationEndpoint) {
    getCloudFoundryInfo(function(error, responseBody) {
    ...
   })
}

requestUserInfoFromUAA(req, callback);

File: api/lib/oauth/oauth.js [hunk 29]
```diff
@@ -116,11 +112,11 @@ module.exports = function(configFilePath) {
...
       }
     };
-    request(options, function(error1, response, body) {
-      if (error1) {
-        logger.error("Failed to get user info from UAA", { "userToken": userToken, "http-options": options, "error": error1 });
-        error1.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
-        callback(error1, null);
+    request(options, function(error, response, body) {
+      if (error) {
+        logger.error("Failed to get user info from UAAss", { "userToken": userToken, "http-options": options, "error": error });
```
Comment (kmj940101) [inline api/lib/oauth/oauth.js]: The error message has a typo, which is "UAAss"
How about using "Failed to get user info from UAA"?

**Comment by cdlliuy** [comment]
LGTM

File: api/test/unit/config/settings_test.js [hunk 30]
```diff
@@ -170,6 +174,40 @@ describe('config setting Test Suite', function() {
...
+      });
+      context('When cfApi is not a string', function() {
+        it('Should return false', function() {
+          settings.cfApi = 12345;
+          expect(settings.validate().valid).to.equal(false);
+          expect(settings.validate().message).to.equal("cfApi must be a string");
+        })
+      });
+      it("Should add http if no protocol",function(){
+        var apiSetting = configSetting({ cfApi: defaultConfig.cfApi.replace("https://",""), db: { uri: defaultConfig.db.uri}, scheduler: { uri: defaultConfig.scheduler.uri, tls: defaultConfig.scheduler.tls }, scalingEngine: { uri: defaultConfig.scalingEngine.uri, tls: defaultConfig.scalingEngine.tls },metricsCollector: { uri: defaultConfig.metricsCollector.uri, tls: defaultConfig.metricsCollector.tls } });
```
Comment (boyang9527) [inline api/test/unit/config/settings_test.js]: why not just set the cfApi to be a test url without scheme? 

The same comments for following similar test cases

File: api/test/unit/oauth/oauth.test.js [hunk 31]
```diff
@@ -0,0 +1,277 @@
...
+      it("should return error", function(done) {
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error.statusCode).to.equal(HttpStatus.INTERNAL_SERVER_ERROR);
+          done();
+        });
+      });
+    });
+
+    context("Cloud Controller /v2/info returns status except 200", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: returns 503

File: api/test/unit/oauth/oauth.test.js [hunk 32]
```diff
@@ -0,0 +1,277 @@
...
+var HttpStatus = require('http-status-codes');
+
+var logger = require("../../../lib/log/logger");
+var oauth;
+var nock = require("nock");
+var theAppId = "the-app-guid";
+var theUserId = "the-user-id";
+var req = {};
+describe("Oauth", function() {
+  this.timeout(10000 + 1000); //timeout is 10s and here wait 11s
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: where the 10s timeout is defined?

File: api/test/unit/oauth/oauth.test.js [hunk 33]
```diff
@@ -0,0 +1,277 @@
...
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error).to.deep.equal({
+            "statusCode": HttpStatus.UNAUTHORIZED
+          });
+          done();
+        });
+      });
+    });
+    context("Authorization endpoint returns status except 200 and 401", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: returns 503

File: api/test/unit/oauth/oauth.test.js [hunk 34]
```diff
@@ -0,0 +1,277 @@
...
+        oauth.checkUserAuthorization(req, function(error, result) {
+          expect(result).to.equal(null);
+          expect(error).to.deep.equal({
+            "statusCode": HttpStatus.UNAUTHORIZED
+          });
+          done();
+        });
+      });
+    });
+    context("Cloud Controller user spaces api returns status except 200 and 401", function() {
```
Comment (boyang9527) [inline api/test/unit/oauth/oauth.test.js]: again returns 503

File: src/integration/integration_api_metricscollector_test.go [hunk 35]
```diff
@@ -53,6 +55,69 @@ var _ = Describe("Integration_Api_MetricsCollector", func() {
...
+				fakeCCNOAAUAA.AllowUnhandledRequests = true
+				fakeCCNOAAUAA.RouteToHandler("GET", "/v2/info", ghttp.RespondWithJSONEncoded(http.StatusOK,
+					cf.Endpoints{
+						AuthEndpoint:    fakeCCNOAAUAA.URL(),
+						DopplerEndpoint: strings.Replace(fakeCCNOAAUAA.URL(), "http", "ws", 1),
+					}))
+				fakeCCNOAAUAA.RouteToHandler("GET", "/userinfo", ghttp.RespondWithJSONEncoded(http.StatusUnauthorized, struct{}{}))
+				parameters = map[string]string{"start-time": "1111", "end-time": "9999", "metric-type": metricType, "order": "asc", "page": "1", "results-per-page": "5"}
+			})
+			It("should error when access public api with status 401", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_metricscollector_test.go]: should error with status code 401

File: src/integration/integration_api_metricscollector_test.go [hunk 36]
```diff
@@ -53,6 +55,69 @@ var _ = Describe("Integration_Api_MetricsCollector", func() {
...
+				fakeCCNOAAUAA.Reset()
+				fakeCCNOAAUAA.AllowUnhandledRequests = true
+				fakeCCNOAAUAA.RouteToHandler("GET", "/v2/info", ghttp.RespondWithJSONEncoded(http.StatusOK,
+					cf.Endpoints{
+						AuthEndpoint:    fakeCCNOAAUAA.URL(),
+						DopplerEndpoint: strings.Replace(fakeCCNOAAUAA.URL(), "http", "ws", 1),
+					}))
+				parameters = map[string]string{"start-time": "1111", "end-time": "9999", "metric-type": metricType, "order": "asc", "page": "1", "results-per-page": "5"}
+			})
+			It("should error when access public api with status 500", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_metricscollector_test.go]: should error with status code 500

File: src/integration/integration_api_scalingengine_test.go [hunk 37]
```diff
@@ -47,6 +48,70 @@ var _ = Describe("Integration_Api_ScalingEngine", func() {
...
+				fakeCCNOAAUAA.AllowUnhandledRequests = true
+				fakeCCNOAAUAA.RouteToHandler("GET", "/v2/info", ghttp.RespondWithJSONEncoded(http.StatusOK,
+					cf.Endpoints{
+						AuthEndpoint:    fakeCCNOAAUAA.URL(),
+						DopplerEndpoint: strings.Replace(fakeCCNOAAUAA.URL(), "http", "ws", 1),
+					}))
+				fakeCCNOAAUAA.RouteToHandler("GET", "/userinfo", ghttp.RespondWithJSONEncoded(http.StatusUnauthorized, struct{}{}))
+				parameters = map[string]string{"start-time": "1111", "end-time": "9999", "order": "desc", "page": "1", "results-per-page": "5"}
+			})
+			It("should error when access public api with status 401", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_scalingengine_test.go]: should error with status code 401

File: src/integration/integration_api_scalingengine_test.go [hunk 38]
```diff
@@ -47,6 +48,70 @@ var _ = Describe("Integration_Api_ScalingEngine", func() {
...
+			BeforeEach(func() {
+				fakeCCNOAAUAA.RouteToHandler("GET", checkUserSpaceRegPath, ghttp.RespondWithJSONEncoded(http.StatusOK,
+					struct {
+						TotalResults int `json:"total_results"`
+					}{
+						0,
+					}))
+				parameters = map[string]string{"start-time": "1111", "end-time": "9999", "order": "desc", "page": "1", "results-per-page": "5"}
+			})
+			It("should error when access public api", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_scalingengine_test.go]: should error with status code 401

File: src/integration/integration_api_scheduler_test.go [hunk 39]
```diff
@@ -38,6 +41,128 @@ var _ = Describe("Integration_Api_Scheduler", func() {
 		stopScheduler(schedulerProcess)
 	})
 
+	Context("Cloud Controller api is not available", func() {
+		BeforeEach(func() {
+			fakeCCNOAAUAA.Reset()
+			fakeCCNOAAUAA.AllowUnhandledRequests = true
+		})
+		Context("Create policy", func() {
+			It("should error when access public api with status 500", func() {
```
Comment (boyang9527) [inline src/integration/integration_api_scheduler_test.go]: should error with status code 500

Comment (boyang9527) [reply src/integration/integration_api_scheduler_test.go]: also the same comments for others

File: api/test/unit/oauth/oauth.test.js (hunk 32 above)
Comment (qibobo) [reply api/test/unit/oauth/oauth.test.js]: apiserver communicate with other components(scalingengine,metricscollector&scheduler), the timeout is 10s. It is defined in lib/utils/*Utils.js

**Comment by qibobo** [comment]
@boyang9527  updated.

File: api/lib/config/setting.js [hunk 40]
```diff
@@ -23,7 +23,7 @@ module.exports = function(settingsObj) {
     };
   var addProtocol = function(uri) {
     if(uri && (uri.indexOf("https://") < 0 && uri.indexOf("http://") < 0)){
-      uri = "https://" + uri;
+      uri = "http://" + uri;
```
Comment (boyang9527) [inline api/lib/config/setting.js]: why change to http

Comment (qibobo) [reply api/lib/config/setting.js]: if protocol is undefined, should it be http or https? How to define?

**Pull request merged**